### ISSUE TITLE
wasm filter config update causes request failing

### ISSUE THREAD

**Issue by bianpengyuan** [istio/istio#29843]
Found this when testing out basic auth wasm filter. When updating Wasm filter configuration (note: only filter configuration is updated, vm config remains the same), envoy starts to return empty response. From the log it looks like Envoy is having problem when reading from the connection:
<details>
<summary>Envoy Trace Log</summary>

```
2021-01-04T05:08:39.698971Z	debug	envoy conn_handler	[C271] new connection
2021-01-04T05:08:39.699018Z	trace	envoy connection	[C271] socket event: 3
2021-01-04T05:08:39.699022Z	trace	envoy connection	[C271] write ready
2021-01-04T05:08:39.699026Z	trace	envoy connection	[C271] read ready. dispatch_buffered_data=false
2021-01-04T05:08:39.699038Z	trace	envoy connection	[C271] read returns: 91
2021-01-04T05:08:39.699049Z	trace	envoy connection	[C271] read error: Resource temporarily unavailable
2021-01-04T05:08:39.699071Z	trace	envoy http	[C271] parsing 91 bytes
2021-01-04T05:08:39.699076Z	trace	envoy http	[C271] message begin
2021-01-04T05:08:39.699081Z	debug	envoy http	[C271] new stream
2021-01-04T05:08:39.699098Z	trace	envoy http	[C271] completed header: key=Host value=34.121.158.42
2021-01-04T05:08:39.699109Z	trace	envoy http	[C271] completed header: key=User-Agent value=curl/7.47.0
2021-01-04T05:08:39.699115Z	trace	envoy http	[C271] onHeadersCompleteBase
2021-01-04T05:08:39.699119Z	trace	envoy http	[C271] completed header: key=Accept value=*/*
2021-01-04T05:08:39.699125Z	trace	envoy http	[C271] Server: onHeadersComplete size=3
2021-01-04T05:08:39.699136Z	trace	envoy http	[C271] message complete
2021-01-04T05:08:39.699143Z	trace	envoy connection	[C271] readDisable: disable=true disable_count=0 state=0 buffer_length=91
2021-01-04T05:08:39.699165Z	debug	envoy http	[C271][S10432564638387055530] request headers complete (end_stream=true):
':authority', 'x.x.x.x'
':path', '/productpage'
':method', 'GET'
'user-agent', 'curl/7.47.0'
'accept', '*/*'

2021-01-04T05:08:39.699200Z	debug	envoy http	[C271][S10432564638387055530] request end stream
2021-01-04T05:08:39.699294Z	debug	envoy http	[C271][S10432564638387055530] doEndStream() resetting stream
2021-01-04T05:08:39.699306Z	debug	envoy http	[C271][S10432564638387055530] stream reset
2021-01-04T05:08:39.699396Z	debug	envoy wasm	wasm log: [extensions/stackdriver/stackdriver.cc:612]::shouldLogThisRequest() cannot get envoy access log info from filter state.
2021-01-04T05:08:39.701671Z	debug	envoy wasm	wasm log: [extensions/stats/plugin.cc:637]::report() metricKey cache miss istio_requests_total , stat=20, recurrent=0
2021-01-04T05:08:39.703668Z	debug	envoy wasm	wasm log: [extensions/stats/plugin.cc:637]::report() metricKey cache miss istio_response_bytes , stat=30, recurrent=0
2021-01-04T05:08:39.707585Z	debug	envoy wasm	wasm log: [extensions/stats/plugin.cc:637]::report() metricKey cache miss istio_request_duration_milliseconds , stat=34, recurrent=0
2021-01-04T05:08:39.711522Z	debug	envoy wasm	wasm log: [extensions/stats/plugin.cc:637]::report() metricKey cache miss istio_request_bytes , stat=38, recurrent=0
2021-01-04T05:08:39.713501Z	trace	envoy main	item added to deferred deletion list (size=1)
2021-01-04T05:08:39.713523Z	trace	envoy connection	[C271] readDisable: disable=false disable_count=1 state=0 buffer_length=91
2021-01-04T05:08:39.713549Z	debug	envoy connection	[C271] closing data_to_write=0 type=2
2021-01-04T05:08:39.713559Z	debug	envoy connection	[C271] setting delayed close timer with timeout 1000 ms
2021-01-04T05:08:39.713567Z	trace	envoy http	[C271][S10432564638387055530] decode headers called: filter=0x55fed1187bc0 status=1
2021-01-04T05:08:39.713575Z	trace	envoy http	[C271] parsed 91 bytes
2021-01-04T05:08:39.713580Z	debug	envoy connection	[C271] closing data_to_write=0 type=2
```

</details>

This happens on istio-ingress and there is only one instance. However this does not happen consistently. Sometime request works fine, which tends to suggest that some workers are not in a good state. I've not yet able to reproduce it locally.

**Comment by bianpengyuan** [comment]
After a config update and envoy gets into a bad state, even after deleting and recreating the filter, the issue still happens.

If the filter config never updates, but only gets deleted and created again, this won't happen. So it looks like config update causes some workers to be stuck at the bad state.

**Comment by bianpengyuan** [comment]
cc @kyessenov @PiotrSikora @mandarjog in case you have suspicion before I spent time digging more into this..

**Comment by kyessenov** [comment]
I don't have any leads. How do we reproduce this? There should be a separate VM for a new config now I think.

**Comment by bianpengyuan** [comment]
> How do we reproduce this?

I've not able to reproduce with locally, will get to that.

> There should be a separate VM for a new config now I think.

You mean even just changing the plugin configuration (VM config remains unchanged), there will be VM creation? That does match my memory..

**Comment by kyessenov** [comment]
> You mean even just changing the plugin configuration (VM config remains unchanged), there will be VM creation? That does match my memory..

This is a new change in envoy. When a module config changes, the old one needs to continue to serve draining listeners. So there is a new VM spawn off keyed by the config. @PiotrSikora to confirm.

**Comment by bianpengyuan** [comment]
Seems like this does not happen any more.

**Comment by bianpengyuan** [comment]
Actually this still happens.. Debugging further on this.

**Comment by PiotrSikora** [comment]
> This is a new change in envoy. When a module config changes, the old one needs to continue to serve draining listeners. So there is a new VM spawn off keyed by the config. @PiotrSikora to confirm.

No, the new change is that we're creating a new root context for each config update instead of updating the same root context in place. It should not affect the VM lifecycle.

**Comment by bianpengyuan** [comment]
Step to reproduce locally: https://github.com/proxy-wasm/proxy-wasm-cpp-host/issues/125#issuecomment-775546438

**Comment by mbfmbf** [comment]
Hello,
after applying the two examples of basic auth manifest, I can authenticate myself on by command line "wget, curl" but not in the browser, how to correct this?

```console
wget -qS https://www.example.dev/productpage
  HTTP/1.1 401 Unauthorized
  date: Mon, 15 Feb 2021 15:50:51 GMT
  server: istio-envoy
  content-length: 0

wget --header "authorization: Basic YWRtaW4zOmFkbWluMw==" -qS https://www.example.dev/productpage
  HTTP/1.1 200 OK
  server: istio-envoy
  date: Mon, 15 Feb 2021 15:51:02 GMT
  content-type: text/html
  content-length: 2344
  last-modified: Thu, 04 Feb 2021 11:21:21 GMT
  vary: Accept-Encoding
  etag: "601bd8b1-928"
  accept-ranges: bytes
  x-envoy-upstream-service-time: 12
```

**Comment by mandarjog** [comment]
WWW-Authenticate header is not being sent.
https://github.com/istio-ecosystem/wasm-extensions/pull/48 fixes it.

**Comment by bianpengyuan** [comment]
I verified this has been fixed after updating envoy sha at 1.9.
